api/all/src/main/java/io/opentelemetry/api/internal/ImmutableKeyValuePairs.java

```diff
@@ -181,31 +182,39 @@ private static void splitAndMerge(
   }
 
   @SuppressWarnings("unchecked")
-  private static <K> List<Object> dedupe(Object[] data, Comparator<K> keyComparator) {
-    List<Object> result = new ArrayList<>(data.length);
+  private static <K> Object[] dedupe(Object[] data, Comparator<K> keyComparator) {
     Object previousKey = null;
+    int size = 0;
 
-    // iterate in reverse, to implement the "last one in wins" behavior.
-    for (int i = data.length - 2; i >= 0; i -= 2) {
+    // Implement the "last one in wins" behavior.
+    for (int i = 0; i < data.length; i += 2) {
       Object key = data[i];
       Object value = data[i + 1];
+      // Skip entries with key null.
       if (key == null) {
         continue;
       }
+      // If the previously added key is equal with the current key, we overwrite what we have.
       if (previousKey != null && keyComparator.compare((K) key, (K) previousKey) == 0) {
-        continue;
+        size -= 2;
       }
-      previousKey = key;
+      // Skip entries with null value, we do it here because we want them to overwrite and remove
+      // entries with same key that we already added.
       if (value == null) {
         continue;
       }
-      // add them in reverse order, because we'll reverse the list before returning,
-      // to preserve insertion order.
-      result.add(value);
-      result.add(key);
+      previousKey = key;
+      data[size++] = key;
+      data[size++] = value;
     }
-    Collections.reverse(result);
-    return result;
+    // Elements removed from the array, copy the array. We optimize for the case where we don't have
+    // duplicates or invalid entries.
+    if (data.length != size) {
```

Comment:
Do you think we'd add `size` as a field in a future PR to avoid this copy?

Reply:
I am working on that, but the problem is with hash/equal which are no longer easy to implement. But yes I think we can do that.